plugin_runner/plugin_installer.py

```diff
@@ -0,0 +1,205 @@
+# Plugin "packages" include this prefix in the database record for the plugin and the S3 bucket key.
+UPLOAD_TO_PREFIX = "plugins"
+
+
+class PluginError(Exception):
+    """An exception raised for plugin-related errors."""
+
+
+class PluginValidationError(PluginError):
+    """An exception raised when a plugin package is not valid."""
+
+
+class InvalidPluginFormat(PluginValidationError):
+    """An exception raised when the plugin file format is not supported."""
+
+
+class PluginInstallationError(PluginError):
+    """An exception raised when a plugin fails to install."""
+
+
+def get_database_dict_from_url() -> dict[str, Any]:
+    """Creates a psycopg ready dictionary from the home-app database URL."""
+    parsed_url = parse.urlparse(os.getenv("DATABASE_URL"))
+    db_name = parsed_url.path[1:]
+    return {
+        "dbname": db_name,
+        "user": parsed_url.username,
+        "password": parsed_url.password,
+        "host": parsed_url.hostname,
+        "port": parsed_url.port,
+    }
+
+
+def get_database_dict_from_env() -> dict[str, Any]:
+    """Creates a psycopg ready dictionary from the environment variables."""
+    return {
+        "dbname": "home-app",
+        "user": os.getenv("DB_USERNAME", "app"),
+        "password": os.getenv("DB_PASSWORD", "app"),
+        "host": os.getenv("DB_HOST", "home-app-db"),
+        "port": os.getenv("DB_PORT", "5432"),
+    }
+
+
+def open_database_connection() -> Connection:
+    """Opens a psycopg connection to the home-app database."""
+    # When running within Aptible, use the database URL, otherwise pull from the environment variables.
+    if os.getenv("DATABASE_URL"):
+        database_dict = get_database_dict_from_url()
+    else:
+        database_dict = get_database_dict_from_env()
+    conn = psycopg.connect(**database_dict)
+    return conn
+
+
+def enabled_plugins() -> dict[str, dict[str, str | dict[str, str]]]:
+    """Returns a dictionary of enabled plugins and their attributes."""
+    conn = open_database_connection()
+
+    plugins = {}
+
+    with conn.cursor() as cursor:
+        cursor.execute(
+            "select name, package, version, key, value from plugin_io_plugin p "
+            "join plugin_io_pluginsecret s on p.id = s.plugin_id where is_enabled"
+        )
+        rows = cursor.fetchall()
+        for row in rows:
+            if row["name"] not in plugins:
+                plugins[row["name"]] = {
+                    "version": row["version"],
+                    "package": row["package"],
+                    "secrets": {row["key"]: row["value"]},
+                }
+            else:
+                plugins[row["name"]]["secrets"][row["key"]] = row["value"]
+
+    return plugins
+
+
+def download_plugin(plugin_package: str) -> Path:
+    """Download the plugin package from the S3 bucket."""
+    s3 = boto3.client("s3")
+    temp_dir = tempfile.TemporaryDirectory()
+    prefix_dir = os.path.join(temp_dir.name, UPLOAD_TO_PREFIX)
```

Comment:
I think you can replace any use of `os.path.join` with use of `Path` and `/`, which is considered the more modern option.

```suggestion
    prefix_dir = Path(temp_dir.name) / UPLOAD_TO_PREFIX
```